File: app/src/main/java/com/robifr/ledger/ui/search_customer/recycler/SearchCustomerAdapter.java
/**
 * Copyright (c) 2022-present Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.search_customer.recycler;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.robifr.ledger.data.model.CustomerModel;
import com.robifr.ledger.databinding.CustomerCardBinding;
import com.robifr.ledger.databinding.ListableListTextBinding;
import com.robifr.ledger.ui.RecyclerViewHolder;
import com.robifr.ledger.ui.search_customer.SearchCustomerFragment;
import com.robifr.ledger.util.Enums;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class SearchCustomerAdapter extends RecyclerView.Adapter<RecyclerViewHolder> {
  private enum ViewType {
    HEADER(0),
    LIST(1);

    private final int _value;

    private ViewType(int value) {
      this._value = value;
    }

    public int value() {
      return this._value;
    }
  }

  @NonNull private final SearchCustomerFragment _fragment;

  public SearchCustomerAdapter(@NonNull SearchCustomerFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
  }

  @Override
  @NonNull
  public RecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    Objects.requireNonNull(parent);

    final ViewType type =
        Objects.requireNonNull(Enums.valueOf(viewType, ViewType.class, ViewType::value));
    final LayoutInflater inflater = this._fragment.getLayoutInflater();

    return switch (type) {
      case HEADER ->
          new SearchCustomerHeaderHolder(
              this._fragment, ListableListTextBinding.inflate(inflater, parent, false));

        // Defaults to `ViewType#LIST`.
      default ->
          new SearchCustomerListHolder(
              this._fragment,
              CustomerCardBinding.inflate(this._fragment.getLayoutInflater(), parent, false));
    };
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerViewHolder holder, int index) {
    Objects.requireNonNull(holder);

    if (holder instanceof SearchCustomerHeaderHolder headerHolder) {
      headerHolder.bind(Optional.empty());

    } else if (holder instanceof SearchCustomerListHolder listHolder) {
      final List<CustomerModel> customers =
          this._fragment.searchCustomerViewModel().customers().getValue();

      // -1 offset because header holder.
      if (customers != null) listHolder.bind(customers.get(index - 1));
    }
  }

  @Override
  public int getItemCount() {
    final List<CustomerModel> customers =
        this._fragment.searchCustomerViewModel().customers().getValue();
    return customers != null ? customers.size() + 1 : 0; // +1 offset because header holder.
  }

  @Override
  public int getItemViewType(int index) {
    return switch (index) {
      case 0 -> ViewType.HEADER.value();
      default -> ViewType.LIST.value();
    };
  }
}

File: app/src/main/java/com/robifr/ledger/ui/searchcustomer/recycler/SearchCustomerAdapter.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.searchcustomer.recycler

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.robifr.ledger.databinding.CustomerCardWideBinding
import com.robifr.ledger.databinding.ListableListTextBinding
import com.robifr.ledger.ui.RecyclerViewHolder
import com.robifr.ledger.ui.customer.recycler.CustomerListHolder
import com.robifr.ledger.ui.searchcustomer.SearchCustomerFragment
import com.robifr.ledger.ui.selectcustomer.recycler.SelectCustomerListHolder

class SearchCustomerAdapter(private val _fragment: SearchCustomerFragment) :
    RecyclerView.Adapter<RecyclerViewHolder>() {
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolder =
      when (ViewType.entries.find { it.value == viewType }) {
        ViewType.HEADER ->
            SearchCustomerHeaderHolder(
                _textBinding =
                    ListableListTextBinding.inflate(_fragment.layoutInflater, parent, false),
                _customers = { _fragment.searchCustomerViewModel.uiState.safeValue.customers })
        else -> {
          val cardBinding: CustomerCardWideBinding =
              CustomerCardWideBinding.inflate(_fragment.layoutInflater, parent, false)
          if (_fragment.searchCustomerViewModel.uiState.safeValue.isSelectionEnabled) {
            SelectCustomerListHolder(
                _cardBinding = cardBinding,
                _initialSelectedCustomerIds = {
                  _fragment.searchCustomerViewModel.uiState.safeValue.initialSelectedCustomerIds
                },
                _customers = { _fragment.searchCustomerViewModel.uiState.safeValue.customers },
                _onCustomerSelected = _fragment.searchCustomerViewModel::onCustomerSelected,
                _expandedCustomerIndex = {
                  _fragment.searchCustomerViewModel.uiState.safeValue.expandedCustomerIndex
                },
                _onExpandedCustomerIndexChanged =
                    _fragment.searchCustomerViewModel::onExpandedCustomerIndexChanged)
          } else {
            CustomerListHolder(
                _cardBinding = cardBinding,
                _customers = { _fragment.searchCustomerViewModel.uiState.safeValue.customers },
                _expandedCustomerIndex = {
                  _fragment.searchCustomerViewModel.uiState.safeValue.expandedCustomerIndex
                },
                _onExpandedCustomerIndexChanged =
                    _fragment.searchCustomerViewModel::onExpandedCustomerIndexChanged,
                _onCustomerMenuDialogShown =
                    _fragment.searchCustomerViewModel::onCustomerMenuDialogShown)
          }
        }
      }

  override fun onBindViewHolder(holder: RecyclerViewHolder, position: Int) {
    when (holder) {
      is SearchCustomerHeaderHolder -> holder.bind()
      is CustomerListHolder -> holder.bind(position - 1)
      is SelectCustomerListHolder -> holder.bind(position - 1)
    }
  }

  override fun getItemCount(): Int =
      // +1 offset because header holder.
      _fragment.searchCustomerViewModel.uiState.safeValue.customers.size + 1

  override fun getItemViewType(position: Int): Int =
      when (position) {
        0 -> ViewType.HEADER.value
        else -> ViewType.LIST.value
      }

  private enum class ViewType(val value: Int) {
    HEADER(0),
    LIST(1)
  }
}
